cpp/src/arrow/compute/exec/query_context.h

```diff
@@ -0,0 +1,173 @@
+            ///
+            /// \param fn The task to run. Takes no arguments and returns a Status.
+            Status ScheduleTask(std::function<Status()> fn);
+            /// \brief Add a single function as a task to the query's task group
+            ///        on the compute threadpool.
+            ///
+            /// \param fn The task to run. Takes the thread index and returns a Status.
+            Status ScheduleTask(std::function<Status(size_t)> fn);
+            /// \brief Add a single function as a task to the query's task group on
+            ///        the IO thread pool
+            ///
+            /// \param fn The task to run. Returns a status.
+            Status ScheduleIOTask(std::function<Status()> fn);
+
+            // Register/Start TaskGroup is a way of performing a "Parallel For" pattern:
+            // - The task function takes the thread index and the index of the task
+            // - The on_finished function takes the thread index
+            // Returns an integer ID that will be used to reference the task group in
+            // StartTaskGroup. At runtime, call StartTaskGroup with the ID and the number of times
+            // you'd like the task to be executed. The need to register a task group before use will
+            // be removed after we rewrite the scheduler.
+            /// \brief Register a "parallel for" task group with the scheduler
+            ///
+            /// \param task The function implementing the task. Takes the thread_index and
+            ///             the task index.
+            /// \param on_finished The function that gets run once all tasks have been completed.
+            /// Takes the thread_index.
+            ///
+            /// Must be called inside of ExecNode::Init.
+            int RegisterTaskGroup(std::function<Status(size_t, int64_t)> task,
+                                  std::function<Status(size_t)> on_finished);
+
+            /// \brief Start the task group with the specified ID. This can only
+            ///        be called once per task_group_id.
+            ///
+            /// \param task_group_id The ID  of the task group to run
+            /// \param num_tasks The number of times to run the task
+            Status StartTaskGroup(int task_group_id, int64_t num_tasks);
+
+            struct TempFileIOMark
+            {
+                QueryContext *ctx_;
+                size_t bytes_;
+                
+                TempFileIOMark(QueryContext *ctx, size_t bytes)
+                    :
+                    ctx_(ctx),
+                    bytes_(bytes)
+                {
+                    ctx_->in_flight_bytes_to_disk_ += bytes_;
+                }
+
+                ~TempFileIOMark()
+                {
+                    ctx_->in_flight_bytes_to_disk_ -= bytes_;
+                }
+            };
+
+            TempFileIOMark ReportTempFileIO(size_t bytes)
+            {
+                return { this, bytes };
+            }
```

Comment:
This could use some comments / explanation.

Reply:
Added a comment.